inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py

```diff
@@ -0,0 +1,363 @@
+        mock_get_path.side_effect = lambda path: path
+
+        if should_raise:
+            with pytest.raises(SourceError):
+                manager.remove(parameters)
+        else:
+            manager.remove(parameters)
+            assert mock_remove_file.call_count == 2  # Both the file and the key should be removed
+
+    @pytest.mark.parametrize('file_name, should_raise', [
+        ('test', False),  # Normal case
+        ('..', True),  # Invalid file name
+        ('.', True),  # Invalid file name
+        ('test/test', True),  # Invalid file name
+    ])
+    def test_remove_invalid_file_name(self, manager, file_name, should_raise):
+        parameters = ApplicationRemoveSourceParameters(source_list_file_name=file_name, gpg_key_name='test_key')
+
+        if should_raise:
+            with pytest.raises(SourceError):
+                manager.remove(parameters)
+        else:
+            manager.remove(parameters)  # Should not raise
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerUpdate
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerUpdate:
+    @pytest.fixture
+    def manager(self):
+        broker = MagicMock()
+        return UbuntuApplicationSourceManager(broker)
+
+    @pytest.fixture
+    def parameters(self):
+        return ApplicationUpdateSourceParameters(source_list_file_name='test_file', sources=['test_source'])
+
+    @pytest.mark.parametrize('exception, expected_exception', [
+        (IOError('test error'), SourceError),
+        (None, None)
+    ])
+    def test_update(self, manager, parameters, exception, expected_exception):
+        # Mock the os.path.join function to return a dummy file path
+        with patch('os.path.join', return_value='dummy_file_path'):
+            # Mock the create_file_with_contents function
+            with patch('inbm_common_lib.utility.create_file_with_contents') as mock_create:
+                # Mock the open function to prevent actual file operations
+                with patch('builtins.open', MagicMock()):
+                    # If an exception is specified, configure the mock to raise it
+                    if exception is not None:
+                        mock_create.side_effect = exception
+
+                    # Call the function
+                    try:
+                        manager.update(parameters)
+                    except SourceError as e:
+                        # If an expected exception is specified, assert that it is raised
+                        if expected_exception is not None:
+                            assert isinstance(e, expected_exception)
+                    else:
+                        # Otherwise, assert that the mock was called with the correct arguments
+                        mock_create.assert_called_once_with('dummy_file_path', ['test_source'])
+
+#DO NOT DELETE THIS LINE - TestUbuntuOsSourceManagerAdd
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuOsSourceManagerAdd:
```

Comment:
To go back to Pull Request Summary table, Click on the [link](https://github.com/ankit-r-gajera/genai_ag_intel-inb-manageability/pull/4#)

<h3>Class: UbuntuOsSourceManager |Function: add</h3>

[Source Code Link](https://github.com/ankit-r-gajera/genai_ag_intel-inb-manageability/blob/NEX_MAXIMUS_UT/develop/inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager.py?plain=1#L40)

[Unit Test Case Code Link](https://github.com/ankit-r-gajera/genai_ag_intel-inb-manageability/blob/NEX_MAXIMUS_UT/develop/inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py?plain=1#L190)

<h4>Test Case Scenarios:</h4>
Here are the test cases generated by the unit test along with their explanations:

1. `test_add_sources`: This test case is parametrized with different lists of sources. It tests the `add` method of the `UbuntuOsSourceManager` class by mocking the `open` function and checking that the sources are correctly written to the file. The test case is run multiple times with different inputs:
    - Adding a single source with a simple URL.
    - Adding multiple sources with simple URLs.
    - Adding a single source with a complex URL.
    - Adding multiple sources with complex URLs.

2. `test_add_sources_os_error`: This test case is parametrized with a single source. It tests the `add` method by mocking the `open` function to raise an `OSError` and checking that a `SourceError` is raised. This test case simulates a situation where there is an error opening the file (for example, if the file does not exist or there are insufficient permissions to open it).

3. `test_add_sources_edge_cases`: This test case is parametrized with different lists of sources that represent edge cases. It tests the `add` method by mocking the `open` function and checking that the sources are correctly written to the file. The test case is run multiple times with different inputs:
    - Adding a source with a URL that points to a local network or localhost.
    - Adding a source with a URL that uses a non-standard protocol.
    - Adding a source with a URL that includes user information.
    - Adding a source with a URL that includes a port number.
    - Adding a source with a URL that includes a query string or fragment.

These test cases cover a wide range of scenarios and edge cases, ensuring that the `add` method works correctly in different situations.


<h4>Test Execution Results:</h4>

| Total Cases    | 13  |
| -------------- | -- |
| Passed Cases   | 13  |
| Failed Cases   | 0  |



To go back to Pull Request Summary table, Click on the [link](https://github.com/ankit-r-gajera/genai_ag_intel-inb-manageability/pull/4#)